website/project/views/contributor.py

```diff
@@ -665,22 +672,26 @@ def claim_user_registered(auth, node, **kwargs):
     }
     session.save()
 
+    # If a user is already validated though external auth, it is OK to claim
+    should_claim = check_external_auth(auth.user)
     form = PasswordForm(request.form)
     if request.method == 'POST':
         if form.validate():
             if current_user.check_password(form.password.data):
-                node.replace_contributor(old=unreg_user, new=current_user)
-                node.save()
-                status.push_status_message(
-                    'You are now a contributor to this project.',
-                    kind='success',
-                    trust=False
-                )
-                return redirect(node.url)
+                should_claim = True
```

Comment:
I think with this you want to put the exit condition first, it's a bit cleaner like this:
```python
    # If a user is already validated though external auth, it is OK to claim
    if check_external_auth(auth.user):
        node.replace_contributor(old=unreg_user, new=current_user)
        node.save()
        status.push_status_message(
            'You are now a contributor to this project.',
            kind='success',
            trust=False
        )
        return redirect(node.url)

    form = PasswordForm(request.form)
    if request.method == 'POST':
        if not form.validate():
            forms.push_errors_to_status(form.errors)
        if not current_user.check_password(form.password.data):
            status.push_status_message(language.LOGIN_FAILED, kind='warning', trust=False)
```

Reply:
If the user isn't externally authenticated they should still go through the existing check on the form - it's only if they're *only* authenticated through ORCID that they get to skip the form entirely. This way seems to assume that external auth is the only path to authentication, when really it's either being externally validated or having an OSF login and still needing to re-enter your password. Repeating the claim auth code in both of those situations seemed redundant hence the variable check.